go-controller/pkg/network-controller-manager/node_network_controller_manager.go

```diff
@@ -0,0 +1,213 @@
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	kexec "k8s.io/utils/exec"
+)
+
+// nodeNetworkControllerManager structure is the object manages all controllers for all networks for ovnkube-node
+type nodeNetworkControllerManager struct {
+	name           string
+	client         clientset.Interface
+	Kube           kube.Interface
+	watchFactory   factory.NodeWatchFactory
+	stopChan       chan struct{}
+	recorder       record.EventRecorder
+	isOvnUpEnabled bool
+
+	defaultNodeNetworkController BaseNetworkController
+
+	// net-attach-def controller handle net-attach-def and create/delete secondary controllers
+	// nil in dpu-host mode
+	nadController *netAttachDefinitionController
+}
+
+// NewNetworkController create secondary node network controllers for the given NetInfo and NetConfInfo
+func (ncm *nodeNetworkControllerManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (NetworkController, error) {
+	topoType := netConfInfo.TopologyType()
+	if topoType == ovntypes.Layer3Topology || topoType == ovntypes.Layer2Topology ||
+		topoType == ovntypes.LocalnetTopology {
+		return node.NewSecondaryNodeNetworkController(ncm.newCommonNetworkControllerInfo(), nInfo, netConfInfo), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topoType)
+}
+
+// CleanupDeletedNetworks cleans up all stale entities giving list of all existing secondary network controllers
+func (ncm *nodeNetworkControllerManager) CleanupDeletedNetworks(allControllers []NetworkController) error {
+	existingNetworksMap := map[string]bool{}
+	for _, nc := range allControllers {
+		existingNetworksMap[nc.GetNetworkName()] = true
+	}
+
+	// find all secondary network OVN-k8s OVS interfaces
+	ovsArgs := []string{"--columns=name,external_ids", "--data=bare", "--no-headings", "--format=csv",
+		"find", "Interface", "external_ids:sandbox!=\"\"",
+		fmt.Sprintf("external_ids:%s!=\"\"", ovntypes.NetworkExternalID)}
+
+	out, stderr, err := util.RunOVSVsctl(ovsArgs...)
+	if err != nil {
+		return fmt.Errorf("failed to list ovn-k8s OVS interfaces:, stderr: %q, error: %v", stderr, err)
+	}
+
+	// delete all OVS interface of the stale networks
+	lines := strings.Split(out, "\n")
+	for _, line := range lines {
+		cols := strings.Split(line, ",")
+		// Note: There are exactly 2 column entries as requested in the ovs query
+		// Col 0: interface name
+		// Col 1: space separated key=val pairs of external_ids attributes
+		if len(cols) < 2 {
+			// unlikely to happen
+			continue
+		}
+		name := strings.TrimSpace(cols[0])
+		for _, attr := range strings.Split(cols[1], " ") {
+			keyVal := strings.SplitN(attr, "=", 2)
+			if len(keyVal) != 2 || keyVal[0] != ovntypes.NetworkExternalID {
+				continue
+			}
+			netName := keyVal[1]
+			if _, ok := existingNetworksMap[netName]; !ok {
+				_, stderr, err := util.RunOVSVsctl("--if-exists", "--with-iface", "del-port", name)
```

Comment:
> This is not specific to localnet so I wonder why we are doing it now.
> But I am also unsure if we should be doing this at all.
> Should we remove the kernel interface as well?

yes, this was not localnet related. but it is to cleanup any OVS port that associated with any secondary network deleted during ovnkube-node  shutdown. kind of like Sync() function in ovnkube-master.

> Also why not follow the same strategy as with master, and create a dummy controller and call Cleanup on it?

we can do it, and just does not seem necessary. in onkbue-master, we do this because Cleanup for different types (localnet, layer2, layer3) of network controller are different, but here the handling are all the same.

Reply:
> yes, this was not localnet related. but it is to cleanup any OVS port that associated with any secondary network deleted during ovnkube-node shutdown. kind of like Sync() function in ovnkube-master.

What about the kernel interface?

> we can do it, and just does not seem necessary. in onkbue-master, we do this because Cleanup for different types (localnet, layer2, layer3) of network controller are different, but here the handling are all the same.

From a design perspective, the node network controller manager does not have to know if they are the same or not. That's not to say that if they are the same, the can't share the code, but not in node network controller manager, maybe in the base node network controller.

Reply:
sure, will fix.